"bytes left unconsumed" error occured on azure-servicebus-2.0.0-PREVIEW-7.jar, but same code works well on azure-servicebus-2.0.0-PREVIEW-5.jar

**Focus on exceptions**
2018-10-31 09:47:40.921  WARN 21888 --- [b9-e3c4d2e0b9c7] c.m.a.s.primitives.MessagingFactory      : Reactor error occured

com.microsoft.azure.servicebus.primitives.ServiceBusException: **bytes left unconsumed**, TrackingId: 037bc9f0-f0d8-43b0-b98a-2129fd9ead00, at: 2018-10-31T09:47:40.917-05:00[America/Chicago]
       at com.microsoft.azure.servicebus.primitives.MessagingFactory$RunReactor.run(MessagingFactory.java:586) ~[azure-servicebus-2.0.0-PREVIEW-7.jar:na]
       at java.lang.Thread.run(Thread.java:745) [na:1.8.0_101]
Caused by: org.apache.qpid.proton.engine.TransportException: bytes left unconsumed
       at org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.unwrapInput(SimpleSslTransportWrapper.java:142) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.process(SimpleSslTransportWrapper.java:361) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.impl.ssl.SslImpl$UnsecureClientAwareTransportWrapper.process(SslImpl.java:144) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.impl.TransportImpl.process(TransportImpl.java:1513) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.IOHandler$1.run(IOHandler.java:233) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:60) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:389) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324) ~[proton-j-0.29.0.jar:na]
       at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292) ~[proton-j-0.29.0.jar:na]
       at com.microsoft.azure.servicebus.primitives.MessagingFactory$RunReactor.run(MessagingFactory.java:564) ~[azure-servicebus-2.0.0-PREVIEW-7.jar:na]
       ... 1 common frames omitted


**Describe the bug**
The most recent maven jar for azure-servicebus 2.0.0-PREVIEW-7 started to produce failures reading from message queue: "com.microsoft.azure.servicebus.primitives.ServiceBusException: bytes left unconsumed".

**To Reproduce**
Here is the code that sets up our connection to the queue:

public void initialize_OutQueuesClient(G2_MessageHandler mmh, String outboundQueueName) throws G2CommunicationsException {
        // Initialize
        log.info("Create OB queue client: " + outboundQueueName);
        try {
            IQueueClient outboundQueueClient = new QueueClient(new ConnectionStringBuilder(connectionString, outboundQueueName), ReceiveMode.PEEKLOCK);
            outboundQueueClientMap.put(outboundQueueName, outboundQueueClient);
            // Register message handler on agent side to receive messages from server:
            mmh.setQueueClient(outboundQueueClient);
            outboundQueueClient.registerMessageHandler(mmh, new MessageHandlerOptions(1, false, Duration.ofMinutes(1)));
            //mmh.setG2CommunicationsServices(this, outboundQueueClient);
        } catch (InterruptedException | ServiceBusException e) {
            log.error("ServiceException Error in MessagingSerivces: " + e.getMessage(), e);
            throw new G2CommunicationsException(e.getMessage(), e, "InitializeMessengerServices");
        }
    }

This exact same code works messages on other queues, and the message that is failing repeatedly for me works fine if I revert back to a previous version "2.0.0-PREVIEW-5" of the com.microsoft.azure jar. 

**Expected behavior**
The message should be normally received.

**Observed behavior**
Raise exception.

**Environment (please complete the following information):**
 - OS: [e.g. iOS]
 - SDK Version : azure-servicebus-2.0.0-PREVIEW-7.jar

**Additional context**
I am testing azure-servicebus-2.0.0-PREVIEW-7.jar. I find that if I set the DataLen = 10000; the registered receiver will stop working. not sure whether it have relation with this issue. Following is my testing code( ------ notice that it is not the code generate "bytes left unconsumed" error ------)

import com.microsoft.azure.servicebus.*;
import com.microsoft.azure.servicebus.primitives.ConnectionStringBuilder;
import com.google.gson.Gson;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;

public final class ServiveBusVersionTest {
	static String ConnectionString = "Endpoint=sb://servicebussha.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=xxxxxxx";
	static String QueueName = "testsha";
	static Gson GSON = new com.google.gson.Gson();
	static int **DataLen** = 10000;
	static int SendTimes = 100;

	public static void main(String[] args) {
		try {
			run();
		} catch (Exception e) {
			System.out.println(e.getMessage());
			e.printStackTrace();
		}
	}

	public static void run() throws Exception {

		QueueClient sbClient = new QueueClient(new ConnectionStringBuilder(ConnectionString, QueueName),
				ReceiveMode.PEEKLOCK);
		registerReceiver(sbClient);
		sendMessagesAsync(sbClient);

        Thread.sleep(1000*60*100);
        sbClient.close();
	}

	static CompletableFuture<Void> sendMessagesAsync(QueueClient sendClient) {
		List<CompletableFuture> tasks = new ArrayList<CompletableFuture>();
		for (int i = 0; i < SendTimes; i++) {
			final String messageId = Integer.toString(i);
			Message message = new Message(repeat('@', DataLen));
			message.setContentType("string");
			message.setLabel("test");
			message.setMessageId(messageId);
			message.setTimeToLive(Duration.ofDays(7));
			System.out.printf("\nMessage sending: Id = %s", message.getMessageId());
			tasks.add(sendClient.sendAsync(message));
		}
		return CompletableFuture.allOf(tasks.toArray(new CompletableFuture<?>[tasks.size()]));
	}

	static void registerReceiver(QueueClient queueClient) throws Exception {
		queueClient.registerMessageHandler(new IMessageHandler() {
			public CompletableFuture<Void> onMessageAsync(IMessage message) {
				System.out.printf("\nReceive message: " + message.toString());
				return CompletableFuture.completedFuture(null);
			}

			public void notifyException(Throwable throwable, ExceptionPhase exceptionPhase) {
				System.out.printf(exceptionPhase + "-" + throwable.getMessage());
			}
		}, new MessageHandlerOptions(10, true, Duration.ofDays(7)), Executors.newFixedThreadPool(10));
	}
	
	public static String repeat(char what, int howmany) {
	    char[] chars = new char[howmany];
	    Arrays.fill(chars, what);
	    return new String(chars);
	}
}

We can steady reproduce this issue now. I am confirming with our customer to see whether we can provide their testing code to you.

I use my queue to try to reproduce this issue. I send the long messages ( see message content in attached file ) with “service bus explorer”. I can steady reproduce this issue with my service bus.

![image](https://user-images.githubusercontent.com/39817657/48328697-34276700-e680-11e8-8dd1-d967e412b954.png)

This is the call stack I find out:

2018-11-09 20:00:41 WARN  c.m.a.s.p.MessagingFactory:577 - UnHandled exception while processing events in reactor:
org.apache.qpid.proton.engine.HandlerException: org.apache.qpid.proton.engine.TransportException: bytes left unconsumed
       at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
       at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
       at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
       at com.microsoft.azure.servicebus.primitives.MessagingFactory$RunReactor.run(MessagingFactory.java:564)
       at java.lang.Thread.run(Unknown Source)
Caused by: org.apache.qpid.proton.engine.TransportException: bytes left unconsumed
       at org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.unwrapInput(SimpleSslTransportWrapper.java:142)
       at org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.process(SimpleSslTransportWrapper.java:361)
       at org.apache.qpid.proton.engine.impl.ssl.SslImpl$UnsecureClientAwareTransportWrapper.process(SslImpl.java:144)
       at org.apache.qpid.proton.engine.impl.TransportImpl.process(TransportImpl.java:1513)
       at org.apache.qpid.proton.reactor.impl.IOHandler$1.run(IOHandler.java:233)
       at org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118)
       at org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:60)
       at org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:389)
       at org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
       at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
       at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
       ... 4 common frames omitted

We can see that the exception throw from org.apache.qpid SDK.
![image](https://user-images.githubusercontent.com/39817657/48328711-46090a00-e680-11e8-9e97-d50c9c9ed089.png)

Then the exception capture by our service bus SDK:
![image](https://user-images.githubusercontent.com/39817657/48328729-59b47080-e680-11e8-9d48-5bb286145898.png)

I think this issue have relation with SSL unwrap the messages and find the mismatch of stream and flag ( capacity? ).
![image](https://user-images.githubusercontent.com/39817657/48328738-689b2300-e680-11e8-8f54-c212aa0ebfab.png)

I changed to “2.0.0-PREVIEW-5” version and this “bytes left unconsumed” issue do not reproduce.

--- a/azure-servicebus/pom.xml
+++ b/azure-servicebus/pom.xml
@@ -3,7 +3,7 @@
 	<description>Java library for Azure Service Bus</description>	
 	<modelVersion>4.0.0</modelVersion>		
 	<artifactId>azure-servicebus</artifactId>	
-	<version>2.0.0-PREVIEW-7</version>
+	<version>${client-current-version}</version>
 	<url>https://github.com/Azure/azure-service-bus-java</url>
 
 	<parent>

--- a/azure-servicebus/src/main/java/com/microsoft/azure/servicebus/amqp/CustomIOHandler.java
+++ b/azure-servicebus/src/main/java/com/microsoft/azure/servicebus/amqp/CustomIOHandler.java
@@ -19,7 +19,6 @@ public class CustomIOHandler extends IOHandler
 		}
 
 		Transport transport = Proton.transport();
-		transport.setMaxFrameSize(AmqpConstants.WEBSOCKET_MAX_FRAME_SIZE);
 		transport.sasl();
 		transport.setEmitFlowEventOnSend(false);
 		transport.bind(connection);

--- a/azure-servicebus/src/main/java/com/microsoft/azure/servicebus/amqp/WebSocketConnectionHandler.java
+++ b/azure-servicebus/src/main/java/com/microsoft/azure/servicebus/amqp/WebSocketConnectionHandler.java
@@ -51,6 +51,6 @@ public class WebSocketConnectionHandler extends ConnectionHandler {
     {
         // This is the current limitation of https://github.com/Azure/qpid-proton-j-extensions
         // once, this library enables larger frames - this property can be removed.
-        return 4 * 1024;
+        return AmqpConstants.WEBSOCKET_MAX_FRAME_SIZE;
     }
 }

--- a/azure-servicebus/src/test/java/com/microsoft/azure/servicebus/QueueSendReceiveTests.java
+++ b/azure-servicebus/src/test/java/com/microsoft/azure/servicebus/QueueSendReceiveTests.java
@@ -42,6 +42,10 @@ public class QueueSendReceiveTests extends SendReceiveTests
         String messageId = UUID.randomUUID().toString();
         Message message = new Message("AMQP message");
         message.setMessageId(messageId);
+        if(this.isEntityPartitioned())
+        {
+        	message.setPartitionKey(messageId);
+        }
         this.sender.send(message, transaction);
 
         this.factory.endTransactionAsync(transaction, true).get();
@@ -63,6 +67,10 @@ public class QueueSendReceiveTests extends SendReceiveTests
         String messageId = UUID.randomUUID().toString();
         Message message = new Message("AMQP message");
         message.setMessageId(messageId);
+        if(this.isEntityPartitioned())
+        {
+        	message.setPartitionKey(messageId);
+        }
         this.sender.send(message, transaction);
 
         this.factory.endTransactionAsync(transaction, false).get();
@@ -208,6 +216,10 @@ public class QueueSendReceiveTests extends SendReceiveTests
 
         TransactionContext transaction = this.factory.startTransactionAsync().get();
         this.receiver.complete(receivedMessage.getLockToken(), transaction);
+        if(this.isEntityPartitioned())
+        {
+        	message2.setPartitionKey(receivedMessage.getPartitionKey());
+        }
         this.sender.send(message2, transaction);
         this.factory.endTransactionAsync(transaction, true).get();
 

--- a/pom.xml
+++ b/pom.xml
@@ -13,7 +13,7 @@
 		<proton-j-version>0.29.0</proton-j-version>
 	  	<junit-version>4.12</junit-version>
 		<slf4j-version>1.7.0</slf4j-version>
-		<client-current-version>2.0.0-PREVIEW-7</client-current-version>
+		<client-current-version>2.0.0-PREVIEW-8-SNAPSHOT</client-current-version>
 		<qpid-proton-j-extensions-version>1.0.0</qpid-proton-j-extensions-version>
 	</properties>
 

